### PR TITLE
3.x: XNIO-376 and XNIO-377

### PR DESCRIPTION
**Pull request by msfm**
JIRA:
https://issues.redhat.com/browse/XNIO-376
https://issues.redhat.com/browse/XNIO-377

PR:
3.x: https://github.com/xnio/xnio/pull/232 (this one)
3.8: https://github.com/xnio/xnio/pull/233
3.7: https://github.com/xnio/xnio/pull/234

### CODE REVIEW

**Comment by carterkozak** [comment]
The linked tickets appear to be private, is there any chance they can be opened up for those of us following along at home?

Apologies if this strategy doesn't work for reasons documented internally, but since xnio creates the ThreadFactory, should input Runnables be wrapped to remove up the selector reference instead of relying on a specific Thread implementation? The reason I ask is that I don't want to prevent testing with [loom](https://cr.openjdk.java.net/~rpressler/loom/loom/sol1_part1.html) once it becomes available.

**Comment by dmlloyd** [comment]
Carter is correct: it should be possible for worker threads to terminate temporary selectors on exit without using JBossThread to do it.  We control the threads and the runnables either way.

**Comment by msfm** [comment]
Thanks for your feedback! I'll try to implement it and update this fix.

**Comment by msfm** [comment]
I've created 2 potential patches with different approaches:

1. https://github.com/xnio/xnio/compare/3.x...msfm:3.x_XNIO-376_XNIO-377_rework
   - Add a new `ThreadFactory` (`CleanableWorkerThreadFactory`) that can execute cleanup tasks on thread exit 
   - Add a new `XnioWorker#addTaskOnThreadExit(Runnable task)` that is similar to `JBossThread#onExit(Runnable task)`. This holds cleanup tasks in ThreadLocal.

2. https://github.com/xnio/xnio/compare/3.x...msfm:3.x_XNIO-376_XNIO-377_rework2
   - Add a new `ThreadFactory` (`CleanableWorkerThreadFactory`) that can execute cleanup tasks on thread exit 
   - However, no new method for adding cleanup tasks. Only executing clearing ThreadLocal holding selector.
   - Move `selectorThreadLocal` in `NioXnio` to a new `XnioSelectorHolder` class and invoke that `clear` method to remove ThreadLocal on the thread exit.

Initially, I thought of the 1st one. I thought this was better because it can be extensible to use for another cleanup task. However, I guess using ThreadLocal like this way might have a risk that becomes a cause of classloader leak issues on the application server. (I guess JBossThread might have a similar risk. Because, if I understand correctly,  JBossThread has a possibility to hold strong references to a Runnable task loaded from application classloader. Please allow me if I'm misunderstanding.)

Then, I thought of the 2nd one. That can not add any other cleanup task but only executes cleanup temporary selector specifically. However, this needs to move selectorThreadLocal to a new XnioSelectorHolder utility class, so I concern that clearing selector can be invoked from other places.

**Comment by dmlloyd** [comment]
Thanks for investigating these options.  I like option 2 overall better, but it requires a small change: we do not want to introduce a public API and we do not want to bind implementations to using `Selector`.  Instead of introducing a public `XnioSelectorHolder` class, instead there should be a new method on `Xnio` like `protected void handleThreadExit() {}` which is called by each thread as it exits.  The `NioXnio` subclass would override this method to clear the thread local and close the selector.

WDYT?

**Comment by msfm** [comment]
Thanks for the review! I've just updated this PR with the option 2 + your suggestion.

File: api/src/main/java/org/xnio/XnioWorker.java [hunk 1]
```diff
@@ -1284,6 +1289,42 @@ public Thread run() {
         }
     }
 
+    class CleanableWorkerThreadFactory implements ThreadFactory {
```
Comment (dmlloyd) [inline api/src/main/java/org/xnio/XnioWorker.java]: This does not need to be a separate thread factory.  This behavior should always be enabled.

File: api/src/main/java/org/xnio/XnioWorker.java [hunk 2]
```diff
@@ -1284,6 +1289,42 @@ public Thread run() {
...
+        public Thread newThread(final Runnable r) {
+            return doPrivileged(new PrivilegedAction<Thread>() {
+                public Thread run() {
+                    final Thread taskThread = new Thread(threadGroup, new Runnable() {
+                        @Override
+                        public void run() {
+                            try {
+                                r.run();
+                            } finally {
+                                msg.tracef("Thread [%s] exiting and executing cleanup tasks", Thread.currentThread().getName());
```
Comment (dmlloyd) [inline api/src/main/java/org/xnio/XnioWorker.java]: Since `handleThreadExit` is already logging this information, this log message is not necessary.

**Review by dmlloyd — COMMENTED** [review]
This looks good except for a couple of comments.  I do not see any reason to make this behavior be optional or configurable.

Thanks!

**Comment by msfm** [comment]
Thanks. As per your review, I've updated this PR.

However, since this is a behavior change (although it's a necessary change for the fix), I would like to add an option (system property switch) to revert to the previous behavior in EAP 7 for the sake. So, if it's PR for 3.7 branch, is it okay to add such a system property switch?

**Comment by dmlloyd** [comment]
I understand your argument.  But it could just as easily be argued that since every bug fix is a behavior change, every bug fix should be switchable - an impractical proposition.  In this case, there is no utility to keeping temporary selectors open beyond the lifespan of their thread, outside of which the selector is inaccessible; such an option will only lead to exceptions.  Therefore it is safe to always enable this behavior unconditionally.

**Comment by msfm** [comment]
Understood. Are you OK with the latest fix that I've updated? If so, I'll push the same fix to other branches.

**Comment by dmlloyd** [comment]
Yes, this looks great.  Thanks.

**Comment by msfm** [comment]
Thanks! I've just pushed the same fix to other branches.

**Comment by dmlloyd** [comment]
Merged manually.  Thanks!

**Pull request closed without merging**